lib/tapioca/gem/listeners/foreign_constants.rb

```diff
@@ -0,0 +1,38 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Tapioca
+  module Gem
+    module Listeners
+      class ForeignConstants < Base
+        extend T::Sig
+
+        include Runtime::Reflection
+
+        private
+
+        sig { override.params(event: ScopeNodeAdded).void }
+        def on_scope(event)
+          mixin = event.constant
+          return if Class === mixin # Classes can't be mixed into other constants
+
+          # There are cases where we want to process constants not declared by the current
+          # gem, i.e. "foreign constant". These are constants defined in another gem to which
+          # this gem is applying a mix-in. This pattern is especially common for gems that add
+          # behavior to Ruby standard library classes by mixing in modules to them.
+          #
+          # The way we identify these "foreign constants" is by asking the mixin tracker which
+          # constants have mixed in the current module that we are handling. We add all the
+          # constants that we discover to the pipeline to be processed.
+          Runtime::Trackers::Mixin.constants_with_mixin(mixin).each do |constant, locations|
+            location = locations.first
+            next if location.nil? || !@pipeline.gem.contains_path?(location)
```

Comment:
I was experiencing an issue where running `tapioca gem` on Shopify core was generating RBIs for ALL mixins, not just mixins performed in the gem currently being compiled. This is my fix for this issue. The mixin tracker keeps track of the location where this mixin took place, and checks that it occurred within the gem before generating RBIs.

Two points:
- Currently, the locations are an array of file paths, so I took the first one here, which I believe should correspond to the "closest" location where the mixin took place? Let me know if that sounds right for now.
- I'm going to create a follow-up PR that changes the notion of what a mixin location is, based on the work @paracycle did in [this branch](https://github.com/Shopify/tapioca/compare/uk-eg-mixin-fix#) -- basically, taking the most recent "require" location as the correct location. I wanted to save that for another PR because this one is already quite big.

Reply:
It is going to change soon, as you said, but up to this point, our thinking was "this is coming from this gem if any of the locations match the gem require path prefix". So we used to have this method that wasn't being used, that we can bring back again for this: https://github.com/Shopify/tapioca/commit/bdf5d45cd74f329fd438dd8a9b54fd7f4205a39d